# Building a model to predict guest ridership

### Given this data, do we improve the fit of our model (R^2)?

In [73]:
# This should all be familiar by now...
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn import linear_model, metrics

import statsmodels.api as sm

%matplotlib inline

bike_data = pd.read_csv('https://github.com/ga-students/DAT-NYC-37/raw/master/lessons/lesson-07/assets/dataset/bikeshare.csv')
bike_data.head(3)

instant      dteday  season  yr  mnth  hr  holiday  weekday  workingday  \
0        1  2011-01-01       1   0     1   0        0        6           0   
1        2  2011-01-01       1   0     1   1        0        6           0   
2        3  2011-01-01       1   0     1   2        0        6           0   

   weathersit  temp   atemp   hum  windspeed  casual  registered  cnt  
0           1  0.24  0.2879  0.81        0.0       3          13   16  
1           1  0.22  0.2727  0.80        0.0       8          32   40  
2           1  0.22  0.2727  0.80        0.0       5          27   32

#### The following function was provided for you, however you should be familiar with each step of this function

In [74]:
from sklearn import feature_selection, linear_model

# From last class...
def get_linear_model_metrics(X, y):
    # 1. Define and configure the model:
    model = linear_model.LinearRegression()
    f, pvals = feature_selection.f_regression(X, y)

    # 2. Fit the model
    model.fit(X, y)

    # 3. Evaluate & score the model
    residuals = (y - model.predict(X)).values

    print 'P values:', pvals
    print 'Coefficients:', model.coef_
    print 'y-intercept:', model.intercept_
    print 'R-Squared:', model.score(X,y)
    print
    
    return model

### Our starting point:

*(This was done in class and provided for you)*

- Select `atemp`, `hum`, and `weather` as features
- `weather` is a 4 level factor (1, 2, 3, 4), so we converted it to a dummy variable and removed `weather_4` (*Review Question: Why did we do this?*)
- `casual` Was defined as our target variable. (The number of riders in a given hour of the day)

### The starter code you were given:

In [75]:
weather = pd.get_dummies(bike_data.weathersit)

#create new names for our new dummy variables
weather.columns = ['weather_' + str(i) for i in weather.columns]

#join those new variables back into the larger dataset
bikemodel_data = bike_data.join(weather)

columns_to_keep = ['hum', 'temp', 'weather_1', 'weather_2', 'weather_3']

X = bikemodel_data[columns_to_keep]  # Our input matrix
y = bikemodel_data['casual'] # our target

get_linear_model_metrics(X, y);

P values: [  0.00000000e+00   0.00000000e+00   3.75616929e-73   3.43170021e-22
   1.57718666e-55]
Coefficients: [ -84.01121684  112.68901765  -24.68489063  -21.00314494  -21.71893628]
y-intercept: 55.8412915804
R-Squared: 0.311934605989



## Starter code result

So far, our baseline `R^2` is **`0.3119`**. How much can we improve upon it?

### A decent starting point:
Intuitively, it seems like `workingday` would be a good a feature to choose since we'd expect more riders on such days.

In [76]:
columns_to_keep = ['hum', 'temp', 'weather_1', 'weather_2', 'weather_3', 'workingday']

X = bikemodel_data[columns_to_keep]  # Our input matrix
y = bikemodel_data['casual'] # our target

get_linear_model_metrics(X, y);

X_sample = X.iloc[:10000]
np.linalg.det(X_sample.T.dot(X_sample))

P values: [  0.00000000e+00   0.00000000e+00   3.75616929e-73   3.43170021e-22
   1.57718666e-55   0.00000000e+00]
Coefficients: [ -84.14725133  117.70452779  -26.28394685  -21.25298575  -20.81182552
  -34.25197022]
y-intercept: 77.85987364
R-Squared: 0.415888672027



4.1418313074193592e+18

#### Great! By adding one feature, we've improved our R^2 to 0.4158

### My Approach:

In the starter code we "One-Hot" encoded (i.e. the "dummy variables") the `weathersit` categorical variable. However, this isn't the only categorical variable.

Other categorical variables:
    - season
    - yr
    - mnth
    - weekday
    - hr
    
Let's try encoding each of these into dummy variables, as we did with `weathersit` above

*Question: Why am I One-Hot encoding `yr`, `mnth`, `weekday`, as if they are categorical?*

In [77]:
# Note: Same exact process as before, but now for the `season`, `weekday`, `hr`, `month` columns
season = pd.get_dummies(bike_data.season)
season.columns = ['season_' + str(i) for i in season.columns]

weekday = pd.get_dummies(bike_data.weekday)
weekday.columns = ['weekday_' + str(i) for i in weekday.columns]

hr = pd.get_dummies(bike_data.hr)
hr.columns = ['hr_' + str(i) for i in hr.columns]

month = pd.get_dummies(bike_data.mnth)
month.columns = ['month_' + str(i) for i in month.columns]

#join those new variables back into the larger dataset
bikemodel_data = pd.concat([bike_data, weather, season, month, hr, weekday], axis=1);

In [78]:
bikemodel_data

instant      dteday  season  yr  mnth  hr  holiday  weekday  \
0            1  2011-01-01       1   0     1   0        0        6   
1            2  2011-01-01       1   0     1   1        0        6   
2            3  2011-01-01       1   0     1   2        0        6   
3            4  2011-01-01       1   0     1   3        0        6   
4            5  2011-01-01       1   0     1   4        0        6   
5            6  2011-01-01       1   0     1   5        0        6   
6            7  2011-01-01       1   0     1   6        0        6   
7            8  2011-01-01       1   0     1   7        0        6   
8            9  2011-01-01       1   0     1   8        0        6   
9           10  2011-01-01       1   0     1   9        0        6   
10          11  2011-01-01       1   0     1  10        0        6   
11          12  2011-01-01       1   0     1  11        0        6   
12          13  2011-01-01       1   0     1  12        0        6   
13          14  2011-01-01       1   0     1  13        0        6   
14          15  2011-01-01       1   0     1  14        0        6   
15          16  2011-01-01       1   0     1  15        0        6   
16          17  2011-01-01       1   0     1  16        0        6   
17          18  2011-01-01       1   0     1  17        0        6   
18          19  2011-01-01       1   0     1  18        0        6   
19          20  2011-01-01       1   0     1  19        0        6   
20          21  2011-01-01       1   0     1  20        0        6   
21          22  2011-01-01       1   0     1  21        0        6   
22          23  2011-01-01       1   0     1  22        0        6   
23          24  2011-01-01       1   0     1  23        0        6   
24          25  2011-01-02       1   0     1   0        0        0   
25          26  2011-01-02       1   0     1   1        0        0   
26          27  2011-01-02       1   0     1   2        0        0   
27          28  2011-01-02       1   0     1   3        0        0   
28          29  2011-01-02       1   0     1   4        0        0   
29          30  2011-01-02       1   0     1   6        0        0   
...        ...         ...     ...  ..   ...  ..      ...      ...   
17349    17350  2012-12-30       1   1    12  18        0        0   
17350    17351  2012-12-30       1   1    12  19        0        0   
17351    17352  2012-12-30       1   1    12  20        0        0   
17352    17353  2012-12-30       1   1    12  21        0        0   
17353    17354  2012-12-30       1   1    12  22        0        0   
17354    17355  2012-12-30       1   1    12  23        0        0   
17355    17356  2012-12-31       1   1    12   0        0        1   
17356    17357  2012-12-31       1   1    12   1        0        1   
17357    17358  2012-12-31       1   1    12   2        0        1   
17358    17359  2012-12-31       1   1    12   3        0        1   
17359    17360  2012-12-31       1   1    12   4        0        1   
17360    17361  2012-12-31       1   1    12   5        0        1   
17361    17362  2012-12-31       1   1    12   6        0        1   
17362    17363  2012-12-31       1   1    12   7        0        1   
17363    17364  2012-12-31       1   1    12   8        0        1   
17364    17365  2012-12-31       1   1    12   9        0        1   
17365    17366  2012-12-31       1   1    12  10        0        1   
17366    17367  2012-12-31       1   1    12  11        0        1   
17367    17368  2012-12-31       1   1    12  12        0        1   
17368    17369  2012-12-31       1   1    12  13        0        1   
17369    17370  2012-12-31       1   1    12  14        0        1   
17370    17371  2012-12-31       1   1    12  15        0        1   
17371    17372  2012-12-31       1   1    12  16        0        1   
17372    17373  2012-12-31       1   1    12  17        0        1   
17373    17374  2012-12-31       1   1    12  18        0        1   
17374    17375  2

Now that I've encoded my categorical variables and added the dummy variables to a copy of my original dataframe (`bikemodel_data`), I'm going to exclude the following columns from the original dataframe to produce the vector X:

<div style="color: red;">
<h3>Important Note:</h3>

<p>
We should always discard one of the dummy variables. Why? Because that's all we need to encode the levels for each category. For instance, the categorical variable `weathersit` had 4 factors: 1, 2, 3, 4 but if we know 3 out of 4 of these values, then the remaining value can be deduced. For example, if I tell you it can ONLY be sunny, rainy, windy, or snowy on a given day and I also tell you that tomorrow's not going to be sunny, rainy, or windy, then what kind of day will it be tomorrow?
</p>
</div>

If this is confusing, think about why we only needed one dummy variable for a boolean variable Size (IsLarge), not two dummy variables (IsSmall and IsLarge).

**<p style="color: darkorange;"> !! Key takeaway: If you have a categorical feature with k levels, you create k-1 dummy variables !!</p>**

**Outcome variables:**

    'casual',       # This is our target variable, which is an outcome and therefore not an independent variable
    'registered',   # Also an outcome, not an independent variable.
    'cnt'           # Also an outcome, not an independent variable (actually casual + registered)
    
**Index/id variables:**

    'instant',    # This is just the row number
    'dteday',     # This is a date column, and is neither categorical nor numerical. Furthermore, this information 
                  #  is already present in the `yr`, `mnth`, `weekday` columns
    
**Redundant/correlated variables**

*Remember! When we create dummy variables we must remove one of them to prevent redundancy in our input data.*

    'temp'        # Strongly correlated with atemp, so we only should keep atemp or temp (discussed in class)
    'weather_4'
    'season_4'   
    'weekday_6'
    'month_11'
    'hr_23'
    
**Encoded variables**

*Think about why we need to remove the following variables as well...*

    'weathersit', 
    'season', 
    'weekday',
    'workingday',
    'hr',
    'mnth'

In [117]:
# Define our input variable, X by excluding the columns listed above from our dataset:
columns_to_drop = [
        'instant',
        'dteday',
        'weather_4',
        'season_4',
        'weekday_6',
        'month_11',
        'hr_0',
        'weathersit', 
        'season', 
        'weekday',
        'workingday',
        'hr',
        'mnth',
        'casual',
        'registered', 
        'temp', 
        'cnt'
    ]

X = bikemodel_data.drop(columns_to_drop, axis=1)

# We've only kept these columns
print "COLUMNS: " + " ".join(X.columns.values)

get_linear_model_metrics(X, y);

COLUMNS: yr holiday atemp hum windspeed weather_1 weather_2 weather_3 season_1 season_2 season_3 month_1 month_2 month_3 month_4 month_5 month_6 month_7 month_8 month_9 month_10 month_12 hr_1 hr_2 hr_3 hr_4 hr_5 hr_6 hr_7 hr_8 hr_9 hr_10 hr_11 hr_12 hr_13 hr_14 hr_15 hr_16 hr_17 hr_18 hr_19 hr_20 hr_21 hr_22 hr_23 weekday_0 weekday_1 weekday_2 weekday_3 weekday_4 weekday_5
P values: [  8.09908774e-080   3.15814032e-005   0.00000000e+000   0.00000000e+000
   8.66781628e-033   3.75616929e-073   3.43170021e-022   1.57718666e-055
   7.90819136e-239   1.70941845e-060   1.20544330e-119   7.20393520e-107
   6.04893370e-081   7.48894210e-006   9.89721401e-008   2.10345324e-034
   2.02298294e-036   1.86526710e-043   6.89088813e-027   1.52784974e-026
   8.70854193e-006   8.98402890e-067   6.64996464e-060   3.26347943e-066
   3.30809960e-073   9.64851198e-080   2.34162115e-081   8.63420127e-070
   2.84021955e-043   5.03259555e-015   7.51024245e-003   1.56865913e-009
   9.55402696e-041   5.4256314

In [120]:
"', '".join(X.columns)

"yr', 'holiday', 'atemp', 'hum', 'windspeed', 'weather_1', 'weather_2', 'weather_3', 'season_1', 'season_2', 'season_3', 'month_1', 'month_2', 'month_3', 'month_4', 'month_5', 'month_6', 'month_7', 'month_8', 'month_9', 'month_10', 'month_12', 'hr_1', 'hr_2', 'hr_3', 'hr_4', 'hr_5', 'hr_6', 'hr_7', 'hr_8', 'hr_9', 'hr_10', 'hr_11', 'hr_12', 'hr_13', 'hr_14', 'hr_15', 'hr_16', 'hr_17', 'hr_18', 'hr_19', 'hr_20', 'hr_21', 'hr_22', 'hr_23', 'weekday_0', 'weekday_1', 'weekday_2', 'weekday_3', 'weekday_4', 'weekday_5"

In [157]:
X_poly = PolynomialFeatures(degree=4).fit_transform(bike_data[['atemp', 'hum', 'windspeed']])

X_sub = X[['yr', 'holiday', 'atemp', 'hum', 'windspeed', 'weather_1', 'weather_2', 'hr_1', 'hr_2', 'hr_3', 'hr_4', 'hr_5', 'hr_6', 'hr_7', 'hr_8', 'hr_9', 'hr_10', 'hr_11', 'hr_12', 'hr_13', 'hr_14', 'hr_15', 'hr_16', 'hr_17', 'hr_18', 'hr_19', 'hr_20', 'hr_21', 'hr_22', 'hr_23', 'weekday_0', 'weekday_1', 'weekday_2', 'weekday_3', 'weekday_4', 'weekday_5']] # , 'month_4', 'month_5', 'month_6', 'month_7', 'month_8', 'month_9', 'month_10']] # 'season_1', 'season_2', 'season_3', 'month_1', 'month_2', 'month_3', 'month_4', 'month_5', 'month_6', 'month_7', 'month_8', 'month_9', 'month_10', 'month_12']]
X_interaction = PolynomialFeatures(degree=2, interaction_only=True).fit_transform(X_sub)

# X_trans = pd.concat([X, pd.DataFrame(X_poly), pd.DataFrame(X_interaction)], axis=1)

# get_linear_model_metrics(X_trans, y);
# get_linear_model_metrics(X_poly, y);
get_linear_model_metrics(X_trans, y);

P values: [  8.09908774e-080   3.15814032e-005   0.00000000e+000   0.00000000e+000
   8.66781628e-033   3.75616929e-073   3.43170021e-022   1.57718666e-055
   7.90819136e-239   1.70941845e-060   1.20544330e-119   7.20393520e-107
   6.04893370e-081   7.48894210e-006   9.89721401e-008   2.10345324e-034
   2.02298294e-036   1.86526710e-043   6.89088813e-027   1.52784974e-026
   8.70854193e-006   8.98402890e-067   6.64996464e-060   3.26347943e-066
   3.30809960e-073   9.64851198e-080   2.34162115e-081   8.63420127e-070
   2.84021955e-043   5.03259555e-015   7.51024245e-003   1.56865913e-009
   9.55402696e-041   5.42563140e-075   1.83858247e-094   6.55214084e-112
   3.08895832e-108   4.17937513e-102   6.03744699e-105   3.54290569e-046
   2.37080976e-013   7.55377998e-001   3.33118416e-005   5.89308930e-014
   1.88193460e-030   2.25058832e-113   7.57296501e-015   1.87083828e-039
   1.44468004e-042   4.75688343e-032   4.03826417e-006               nan
   0.00000000e+000   0.00000000e+000   8.

In [98]:
columns_to_drop = [
        'instant',
        'dteday',
        'weathersit', 
        'season', 
        'weekday',
        'workingday',
        'yr',
        'holiday',
        'casual',
        'registered', 
        'temp', 
        'cnt'
    ]

X = bike_data.drop(columns_to_drop, axis=1)

X_poly = PolynomialFeatures().fit_transform(X)

# get_linear_model_metrics(X, y);
# X.head()

pd.DataFrame(X_poly)

get_linear_model_metrics(X_poly, y);

P values: [              nan   6.01115086e-057   8.09908774e-080   1.64546575e-019
   0.00000000e+000   3.15814032e-005   1.59971276e-005   0.00000000e+000
   4.52501205e-091   0.00000000e+000   0.00000000e+000   8.66781628e-033
   2.36330684e-021   1.54405782e-113   1.72124996e-010   0.00000000e+000
   2.66831735e-010   2.65587464e-034   4.89128348e-151   5.45968651e-004
   0.00000000e+000   4.21600599e-056   2.04383955e-110   8.09908774e-080
   1.31417853e-079   6.86616646e-301   2.98526235e-005   5.74513546e-078
   4.66203009e-019   4.01106402e-013   0.00000000e+000   3.03422600e-001
   2.09489758e-110   5.48093114e-001   2.46757110e-271   7.74373649e-007
   5.16390189e-017   2.07802344e-147   2.36953809e-009   0.00000000e+000
   8.81364537e-066   9.57161472e-065   2.53927173e-163   3.28900384e-013
   3.03477255e-193   1.19082795e-017   3.52808551e-057   0.00000000e+000
   6.43793722e-049   3.00544549e-205   3.15814032e-005   7.67235652e-002
               nan   3.69525449e-003   1.

#### Our R^2 increased to `0.588`!

### Note the following conditions hold:

- We have small p-values for all  features
- None of our coefficients have extremely large values nor are they close to zero.

#### However! We haven't yet validated our model on data we haven't seen.

In [84]:
from sklearn.preprocessing import PolynomialFeatures

X_poly = PolynomialFeatures(degree=2).fit_transform(X)

In [87]:
get_linear_model_metrics(X_poly, y);

P values: [             nan   8.09908774e-80   3.15814032e-05 ...,   4.75688343e-32
              nan   4.03826417e-06]
Coefficients: [  0.00000000e+000   1.13308148e+013   1.38295911e+010 ...,
   1.87536742e+011   1.95283573e-145   2.98787974e+010]
y-intercept: 2.12703798292e+12
R-Squared: 0.861728360196



### Questions:

1. Our R^2 ALWAYS increases when we add an input feature. Why? Is this necessarily a good thing? Why or why not?
2. Rather than dropping our original `season` column, try keeping it in the input dataframe X by removing `'season'` from our `columns_to_exclude` list. What happens to our coefficients and why does this happen?
3. Do the same thing for the `workingday`. Again explain what's causing any anomalies you see in our input dataframe X.
4. It can be argued that `yr`, `mnth`, `weekday`, and `hr` are numerical variables, not categorical. Why did we treat them as categorical and are there any drawbacks to this approach?
5. Our input vector `X` has 51 columns! Suppose we could only keep 6 columns in our input. Which columns would you keep and why?

## Additional resources:

Feature selection: http://blog.datadive.net/selecting-good-features-part-i-univariate-selection/